### PR TITLE
[dynamo] Add a lint rule to restrict what 3P library one can import

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143312

As title, this patch prevents developers from importing third party
libraries to patch things in Dynamo, unless there's no other easy
workaround (in which case one would add the library to the allowlist in
`import_linter.py`, as instructed by the lint error).

For instance, if we remove `einops` from the allowlist, we'd get this
```verbatim
>>> Lint for torch/_dynamo/decorators.py:

  Error (IMPORT) Disallowed import

    importing from einops is not allowed, if you believe there's a valid
    reason, please add it to import_linter.py

        608  |# Note: this carefully avoids eagerly import einops.
        609  |# TODO: we should delete this whole _allow_in_graph_einops logic by approximately 2024 Q2
        610  |def _allow_in_graph_einops():
    >>> 611  |    import einops
        612  |
        613  |    try:
        614  |        # requires einops > 0.6.1, torch >= 2.0

  Error (IMPORT) Disallowed import

    importing from einops is not allowed, if you believe there's a valid
    reason, please add it to import_linter.py

        612  |
        613  |    try:
        614  |        # requires einops > 0.6.1, torch >= 2.0
    >>> 615  |        from einops._torch_specific import (  # type: ignore[attr-defined]  # noqa: F401
        616  |            _ops_were_registered_in_torchdynamo,
        617  |        )
        618  |
```

### CODE REVIEW

File: tools/linter/adapters/import_linter.py [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+# Add the allowed third party libraries.
+_MODULE_NAME_ALLOW_LIST.update(
+    [
+        "torch",
+        "sympy",
+        "torch_xla",
+        "_pytest",
+        "functorch",
+        "einops",
+        "the",
```
Comment (zou3519) [inline tools/linter/adapters/import_linter.py:325]: I don't see any "import the" statements in torch/_dynamo -- how did this come up?

File: tools/linter/adapters/import_linter.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+                    line=line_number,
+                    char=None,
+                    code="IMPORT",
+                    severity=LintSeverity.ERROR,
+                    name="Disallowed import",
+                    original=None,
+                    replacement=None,
+                    description=f"""
+importing from {module_name} is not allowed, if you believe there's a valid
+reason, please add it to _MODULE_NAME_ALLOW_LIST in {CURRENT_FILE_NAME}
```
Comment (zou3519) [inline tools/linter/adapters/import_linter.py:364]: Nit: let's state the perils of importing a third-party module. "Please do not import third-party modules in PyTorch unless they're explicitly requirements of PyTorch. Imports of a third-party library may have side effects and other unintentional behavior".

File: tools/linter/adapters/import_linter.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+        "torch_xla",
+        "_pytest",
+        "functorch",
+        "einops",
+        "the",
+        "libfb",
+        "typing_extensions",
+        "triton",
+        "numpy",
+        "torchrec",
```
Comment (zou3519) [inline tools/linter/adapters/import_linter.py:330]: note to self: this should also be excised...

File: tools/linter/adapters/import_linter.py (hunk 2 above)
Comment (zou3519) [reply tools/linter/adapters/import_linter.py:364]: We can also state the remedy. "If you're just checking if a module exists, use sys.modules.get("torchrec")"

File: tools/linter/adapters/import_linter.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+            "xxsubtype",
+            "zipapp",
+            "zipfile",
+            "zipimport",
+            "zlib",
+            "zoneinfo",
+        ]
+    )
+
+# Add the allowed third party libraries.
```
Comment (zou3519) [inline tools/linter/adapters/import_linter.py:316]: Nit: add some more information about why not to add stuff here

File: tools/linter/adapters/import_linter.py (hunk 1 above)
Comment (zou3519) [reply tools/linter/adapters/import_linter.py:325]: Oh lol is this literally "from the" in comments?

Comment (StrongerXi) [reply tools/linter/adapters/import_linter.py:325]: LOL yep: https://github.com/pytorch/pytorch/blob/fda43c98d1bc82eeb493f6e79a1fd5f1636474a5/torch/_dynamo/resume_execution.py#L465

Ughhh I probably have to replace line-by-line regex with some AST-aware thing, but thankfully the linter does seem to provide parsing and processing for that, lemme look more.

Comment (zou3519) [reply tools/linter/adapters/import_linter.py:325]: Maybe just a "from foo import y" regex

Comment (StrongerXi) [reply tools/linter/adapters/import_linter.py:325]: That'll still get picked up in a multi-line string... But maybe suffices in practice, I'll make some decision based on complexity of the alternative.

File: tools/linter/adapters/import_linter.py (hunk 2 above)
Comment (StrongerXi) [reply tools/linter/adapters/import_linter.py:364]: Updated.

File: tools/linter/adapters/import_linter.py (hunk 4 above)
Comment (StrongerXi) [reply tools/linter/adapters/import_linter.py:316]: Updated.

File: tools/linter/adapters/import_linter.py (hunk 1 above)
Comment (StrongerXi) [reply tools/linter/adapters/import_linter.py:325]: Ended up using `_linter.PythonFile` and some basic token checks, quite simple.

File: requirements.txt [hunk 5]
```diff
@@ -19,3 +19,4 @@ ninja
 packaging
 optree>=0.13.0
 cmake
+stdlib_list
```
Comment (zou3519) [inline requirements.txt:22]: I thought we weren't going to take a dependency on stdlib_list?

**Review by zou3519 — APPROVED** [review]
Looks good but please remove the stdlib_list dependency before landing

File: tools/linter/adapters/import_linter.py (hunk 1 above)
Comment (zou3519) [reply tools/linter/adapters/import_linter.py]: Nice!

File: requirements.txt (hunk 5 above)
Comment (StrongerXi) [reply requirements.txt:22]: Oops forgot to remove, updated.

**Comment by StrongerXi** [comment]
Idk why docker builds started to fail with `RuntimeError: operator torchvision::nms does not exist`, rebasing.

**Pull request closed without merging**